### PR TITLE
Solution Smoothing

### PR DESCRIPTION
**Pull request by SeaVegetable**
Add a solution smoothing tool for Cauchy stress.

### CODE REVIEW

File: tools/solution_smooth/src/VisDataPrep_Smooth_Vol.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#include "VisDataPrep_Smooth_Vol.hpp"
+
+VisDataPrep_Smooth_Vol::VisDataPrep_Smooth_Vol()
+{
+  // Data to be written
+  arrayCompSize = 1;
+
+  arrayNames.push_back("CauchyStress");
+  arraySizes.push_back(6);
```
Comment (Goldeninscriptions) [inline tools/solution_smooth/src/VisDataPrep_Smooth_Vol.cpp:9]: 冗余

File: include/MaterialModel_Linear_Elasticity.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// Author: Xinhai Yue
+// ==================================================================
+#include "IMaterialModel.hpp"
+#include "Math_Tools.hpp"
+
+class MaterialModel_Linear_Elasticity : public IMaterialModel
+{
+  public:
+    MaterialModel_Linear_Elasticity(
+       const double &in_module, const double &in_nu );
```
Comment (ju-liu) [inline include/MaterialModel_Linear_Elasticity.hpp:19]: in_module ===> in_E

File: include/MaterialModel_Linear_Elasticity.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    virtual double get_elastic_nu() const {return nu;}
+
+    virtual double get_elastic_mu() const {return mu;}
+
+    virtual double get_elastic_lambda() const {return lambda;}
+
+    virtual Tensor2_3D get_Cauchy_stress( const Tensor2_3D &F ) const;
+
+  private:
+    double modulus_E, nu, lambda, mu;
```
Comment (ju-liu) [inline include/MaterialModel_Linear_Elasticity.hpp:59]: only have modulus_E and nu as data. The rest can be computed.

File: src/Model/MaterialModel_Linear_Elasticity.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    h5w -> write_doubleScalar("mu", mu);
+
+    delete h5w; H5Fclose(file_id);
+  }
+
+  MPI_Barrier(PETSC_COMM_WORLD);
+}
+
+Tensor2_3D MaterialModel_Linear_Elasticity::get_Cauchy_stress( const Tensor2_3D &F ) const
+{
```
Comment (ju-liu) [inline src/Model/MaterialModel_Linear_Elasticity.cpp:61]: Simplify the code.

File: tools/solution_smoothing/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+# binary location
+SET(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR})
+
+INCLUDE(../../conf/basic_variable_log.cmake)
+INCLUDE(../../conf/system_lib_loading.cmake)
+
+SET( perigee_source ${perigee_SOURCE_DIR}/../../src )
+# 1. Preprocessor source cpp
+SET( perigee_analysis_lib_src
```
Comment (ju-liu) [inline tools/solution_smoothing/CMakeLists.txt:32]: minimize dependency.

File: tools/solution_smoothing/driver.cpp [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  std::string part_file("./apart/part");
+
+  int time_start = 0, time_step = 1, time_end = 1;
+  
+  // Input and output files' names
+  std::string isol_bname = "SOL_disp_";
+  std::string osol_bname = "SOL_Cauchy_";
+
+  // Input and output solutions' degrees of freedom
+  int isol_dof = 3;
```
Comment (ju-liu) [inline tools/solution_smoothing/driver.cpp:41]: conexpr int isol_dof = 3;

File: tools/solution_smoothing/include/PDNSolution_Smooth_Stress.hpp
Comment (ju-liu) [inline tools/solution_smoothing/include/PDNSolution_Smooth_Stress.hpp:1]: 回避两个solution类

File: tools/solution_smoothing/src/PLocAssem_Smooth_Vol.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  {
+    double ux_x = 0.0, ux_y = 0.0, ux_z = 0.0;
+    double uy_x = 0.0, uy_y = 0.0, uy_z = 0.0;
+    double uz_x = 0.0, uz_y = 0.0, uz_z = 0.0;
+
+    element->get_R_gradR( qua, &R[0], &dR_dx[0], &dR_dy[0], &dR_dz[0] );
+    
+    for(int ii=0; ii<nLocBas; ++ii)
+    {
+      const int iiisol = isol_dof * ii;
```
Comment (ju-liu) [inline tools/solution_smoothing/src/PLocAssem_Smooth_Vol.cpp:69]: offset

File: tools/solution_smoothing/src/VTK_Writer_Smooth_Vol.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    // Interpolate velocity vector
+    int asize = vdata_ptr->get_arraySizes(0);
+    for(int jj=0; jj<nLocBas; ++jj)
+    {
+      int pt_index = IEN_e[jj];
+      for(int kk=0; kk<asize; ++kk)
+        inputInfo.push_back( pointArrays[0][pt_index * asize + kk ] );
+    }
+    intep.interpolateVTKData( asize, &IEN_e[0], &inputInfo[0],
+        elemptr, dataVecs[0] );
```
Comment (ju-liu) [inline tools/solution_smoothing/src/VTK_Writer_Smooth_Vol.cpp:77]: make it a single line

File: tools/solution_smoothing/vis_smooth.cpp [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+#include "VisDataPrep_Smooth_Vol.hpp"
+#include "VTK_Writer_Smooth_Vol.hpp"
+
+int main( int argc, char * argv[] ) 
+{ 
+  const std::string element_part_file = "epart.h5";
+  const std::string anode_mapping_file = "node_mapping.h5";
+  const std::string pnode_mapping_file = "post_node_mapping.h5";
+  const std::string part_file="./ppart/part";
+  constexpr int dof = 6;
```
Comment (ju-liu) [inline tools/solution_smoothing/vis_smooth.cpp:27]: get rid of this data.

File: tools/solution_smoothing/include/PLocAssem_Smooth_Vol.hpp [hunk 10]
```diff
@@ -0,0 +1,62 @@
+#ifndef PLOCASSEM_SMOOTH_VOL_HPP
+#define PLOCASSEM_SMOOTH_VOL_HPP
```
Comment (ju-liu) [inline tools/solution_smoothing/include/PLocAssem_Smooth_Vol.hpp:2]: 把文件名出现 Smooth_Vol 的地方 改成 Stress_Recovery @SeaVegetable

File: include/MaterialModel_Linear_Elasticity.hpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        const char * const &fname = "material_model.h5" );
+    
+    virtual ~MaterialModel_Linear_Elasticity();
+
+    virtual void print_info() const;
+
+    virtual std::string get_model_name() const
+    {
+      const std::string mname = "Linear_Elasticity";
+      return mname;
```
Comment (ju-liu) [inline include/MaterialModel_Linear_Elasticity.hpp:31]: 考虑写作 return std::string("Linear_Elasticity");

https://stackoverflow.com/questions/3976935/best-way-to-return-an-stdstring-that-local-to-a-function

File: tools/solution_smoothing/include/PLocAssem_Smooth_Vol.hpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+// ============================================================================
+// PLocAssem_Smooth_Vol.hpp
+//
+// Parallel Local Assembly routine for solution smoother
+//
```
Comment (ju-liu) [inline tools/solution_smoothing/include/PLocAssem_Smooth_Vol.hpp:7]: 把使用的技术对应的文献记录在这里。我记得是oden 1971年的IJNME论文？

File: include/MaterialModel_Linear_Elasticity.hpp [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    // Read access to parameters
+    virtual double get_elastic_E() const {return modulus_E;}
+
+    virtual double get_elastic_nu() const {return nu;}
+
+    virtual double get_elastic_mu() const {return 0.5 * modulus_E / (1.0 + nu);}
+
+    virtual double get_elastic_lambda() const {return nu * modulus_E / ((1.0 + nu) * (1.0 - 2.0 * nu));}
+
+    virtual Tensor2_3D get_Cauchy_stress( const Tensor2_3D &F ) const;
```
Comment (ju-liu) [inline include/MaterialModel_Linear_Elasticity.hpp:55]: add comments here

File: src/Model/MaterialModel_Linear_Elasticity.cpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
+#include "MaterialModel_Linear_Elasticity.hpp"
+
+MaterialModel_Linear_Elasticity::MaterialModel_Linear_Elasticity(
+    const double &in_modulus_E, const double &in_nu )
+: modulus_E( in_modulus_E ), nu( in_nu )
+{
+}
```
Comment (ju-liu) [inline src/Model/MaterialModel_Linear_Elasticity.cpp:7]: simpler way.

File: src/Model/MaterialModel_Linear_Elasticity.cpp [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+     "Error: MaterialModel_Linear_Elasticity constructor does not match h5 file.\n" );
+
+  modulus_E = h5r -> read_doubleScalar("/", "modulus_E");
+  nu        = h5r -> read_doubleScalar("/", "nu");
+  
+  delete h5r; H5Fclose(h5file);
+}
+
+MaterialModel_Linear_Elasticity::~MaterialModel_Linear_Elasticity()
+{}
```
Comment (ju-liu) [inline src/Model/MaterialModel_Linear_Elasticity.cpp:26]: default

File: tools/solution_smoothing/src/VTK_Writer_Stress_Recovery.cpp [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+    fnode_ptr -> get_ctrlPts_xyz(nLocBas, &IEN_e[0], &ectrl_x[0], &ectrl_y[0], &ectrl_z[0]);
+
+    elemptr->buildBasis( quad, &ectrl_x[0], &ectrl_y[0], &ectrl_z[0] );
+    
+    // Interpolate nodal coordinates
+    intep.interpolateVTKPts(&IEN_e[0], &ectrl_x[0], &ectrl_y[0], &ectrl_z[0],
+        elemptr, points );
+  
+    std::vector<double> inputInfo; inputInfo.clear();
```
Comment (ju-liu) [inline tools/solution_smoothing/src/VTK_Writer_Stress_Recovery.cpp:66]: {} initialization?

File: src/Model/MaterialModel_Linear_Elasticity.cpp (hunk 14 above)
Comment (SeaVegetable) [reply src/Model/MaterialModel_Linear_Elasticity.cpp]: This constructor has been moved to the header file.

**Review by ju-liu — APPROVED** [review]

**Pull request merged**